substrate/frame/staking/ah-client/src/lib.rs

```diff
@@ -0,0 +1,318 @@
+				message,
+			) {
+				log::error!(target: LOG_TARGET, "Sending `EndSession` to AssetHub failed: {:?}", err);
+			}
+		}
+
+		fn start_session(session_index: u32) {
+			// TODO: pass active validator set somehow???
+
+			let message = Xcm(vec![
+				Instruction::UnpaidExecution {
+					weight_limit: WeightLimit::Unlimited,
+					check_origin: None,
+				},
+				mk_asset_hub_call::<T>(StakingCalls::StartSession(session_index)),
+			]);
+			if let Err(err) = send_xcm::<T::SendXcm>(
+				Location::new(0, [Junction::Parachain(T::AssetHubId::get())]),
+				message,
+			) {
+				log::error!(target: LOG_TARGET, "Sending `StartSession` to AssetHub failed: {:?}", err);
+			}
+		}
+	}
+
+	impl<T> pallet_authorship::EventHandler<T::AccountId, BlockNumberFor<T>> for Pallet<T>
+	where
+		T: Config + pallet_authorship::Config + pallet_session::Config + crate::Config,
+		T::AccountId: Into<AccountId32>,
+	{
+		// Notes the authored block in `BlockAuthors`.
+		fn note_author(author: T::AccountId) {
+			BlockAuthors::<T>::mutate(author.into(), |block_count| {
+				*block_count += 1;
+			});
+		}
+	}
+
+	impl<T: Config>
+		OnOffenceHandler<T::AccountId, pallet_session::historical::IdentificationTuple<T>, Weight>
+		for Pallet<T>
+	where
+		T: pallet_session::Config<ValidatorId = <T as frame_system::Config>::AccountId>,
+		T: pallet_session::historical::Config<
+			FullIdentification = Exposure<<T as frame_system::Config>::AccountId, BalanceOf<T>>,
+			FullIdentificationOf = ExposureOf<T>,
+		>,
+		T::SessionHandler: pallet_session::SessionHandler<<T as frame_system::Config>::AccountId>,
+		T::SessionManager: pallet_session::SessionManager<<T as frame_system::Config>::AccountId>,
+		T::ValidatorIdOf: Convert<
+			<T as frame_system::Config>::AccountId,
+			Option<<T as frame_system::Config>::AccountId>,
+		>,
+		T::AccountId: Into<AccountId32>,
+	{
+		fn on_offence(
+			offenders: &[OffenceDetails<
+				T::AccountId,
+				pallet_session::historical::IdentificationTuple<T>,
+			>],
+			slash_fraction: &[Perbill],
+			slash_session: SessionIndex,
+		) -> Weight {
+			let offenders_and_slashes = offenders
+				.iter()
+				.cloned()
+				.zip(slash_fraction)
+				.map(|(offence, fraction)| Offence {
+					offender: offence.offender.0.into(),
+					reporters: offence.reporters.into_iter().map(|r| r.into()).collect(),
+					slash_fraction: *fraction,
+				})
+				.collect::<Vec<_>>();
+
+			// send the offender immediately over xcm
+			let message = Xcm(vec![
+				Instruction::UnpaidExecution {
+					weight_limit: WeightLimit::Unlimited,
+					check_origin: None,
+				},
+				mk_asset_hub_call::<T>(StakingCalls::NewOffence(
+					slash_session,
+					offenders_and_slashes,
+				)),
+			]);
+			if let Err(err) = send_xcm::<T::SendXcm>(
+				Location::new(0, [Junction::Parachain(T::AssetHubId::get())]),
+				message,
+			) {
+				log::error!(target: LOG_TARGET, "Sending `StartSession` to AssetHub failed: {:?}",
+			err);
+			}
+
+			Weight::zero() // TODO: how much?
```

Comment:
I don't know what to return here. Does this function even consume any weight on the relay chain?

Reply:
Yes, these are executed as part of system tasks. Ideally need to bench these (though this is in relay which is forgiving in terms of weight limits).